harness/determined/core/_checkpoint.py

```diff
@@ -34,6 +34,143 @@ class DownloadMode(enum.Enum):
+        key_ranks[key]["_ranks_"].append(rank)
+
+        full_key = f"{prev_key}/{key}"
+
+        # Merge metadata. We consider two possibilities:
+        # 1. metadata is dictionary AND :
+        #   * if key is not in merged, then recursively unroll and merge metadata; OR
+        #   * if key is in merged, and metadata and merged[key] are dictionaries, merge recursively.
+        # 2. otherwise:
+        #   * key is in merged, then updated key_conflicts and don't merge,
+        #   * key is not in merged, then add key and metadata to merged.
+        if (key not in merged and isinstance(metadata, dict)) or (
+            key in merged and isinstance(metadata, dict) and isinstance(merged[key], dict)
+        ):
+            merged[key], key_ranks[key] = _merge_metadata(
+                {} if key not in merged else merged[key],
+                metadata,
+                rank,
+                key_ranks[key],
+                key_conflicts,
+                full_key,
+            )
+        else:
+            if key in merged:
+                if full_key not in key_conflicts:
+                    key_conflicts[full_key] = key_ranks[key]["_ranks_"].copy()
+                else:
+                    key_conflicts[full_key].append(rank)
+            else:
+                merged[key] = metadata
+
+    return merged, key_ranks
+
+
+def merge_metadata(
+    all_metadata: List[Dict[str, Any]]
+) -> Tuple[Dict[str, Any], Dict[str, List[int]]]:
+    """
+    Given a list of metadata, return:
+      - merged metadata
+      - a dict mapping conflicting keys to ranks that would upload them
+
+    Merging scenarios:
+      - merge dictionaries under the same key
+      - report conflicts for any repeated key that is not a dictionary
+
+    Examples:
+    #  Metadata 1     | Metadata 2       | Conflict |  Result
+    # ----------------|------------------|----------|-----------------
+    #  a: 1           | b: 2             | no       | {a: 1, b: 2}
+    #  a: 1           | a: 1             | yes      | n/a
+    #  a: []          | a: []            | yes      | n/a
+    #  a: []          | a: 1             | yes      | n/a
+    #  a: []          | a: {c: 1}        | yes      | n/a
+    #  a: {c: 1}      | a: {d: 1}        | no       | a: {c: 1, d: 1}
+    #  a: {c: 1}      | a: 1             | yes      | n/a
+    """
+
+    # Stores merged metadata.
+    merged: Dict[str, Any] = {}
+
+    # Maps keys to reporting ranks. It has the same structure as merged with each key
+    # having a list of ranks that reports it. Helps to detect conflicts.
+    key_ranks: Dict[str, Any] = {}
+
+    # Maps fullpath keys (from root) to the list of rank reporting it.
+    key_conflicts: Dict[str, List[int]] = {}
+
+    for rank, rank_metadata in enumerate(all_metadata):
+        merged, key_ranks = _merge_metadata(
+            merged, rank_metadata, rank, key_ranks, key_conflicts, ""
+        )
+
+    return merged, key_conflicts
+
+
+def merge_resources(
```

Comment:
unfortunately deepspeed checkpoint can replicates some utility files across ranks with the same name in certain situations.  this will cause a failed upload.  the replicated files are the same so it's safe to just keep one copy.  i'm not sure what we should do here without making it really messy to use our sharded checkpoint with deepspeed.  we can either just take rank 0 file when there is conflict or do something more intelligent and check whether conflicted files are equal or not, or maybe something else i haven't thought of.